FormalBook/Rainbow_triangles.lean

```diff
@@ -0,0 +1,514 @@
+import Mathlib.Tactic
+import FormalBook.Appendix
+import FormalBook.sperner
+
+
+
+/-!
+# One square and an odd number of triangles
+-/
+
+local notation "ℝ²" => EuclideanSpace ℝ (Fin 2)
+local notation "ℝ³" => EuclideanSpace ℝ (Fin 3)
+local notation "Triangle" => Fin 3 → ℝ²
+local notation "Segment" => Fin 2 → ℝ²
+
+open BigOperators
+open Classical
+open Finset
+
+
+
+
+
+-- First we define the inductive type Rainbow, which will be the the target type of the painter
+-- function. The painter function will take a point in ℝ² and return a color from Rainbow (eg. Red
+-- Blue or Green).
+
+inductive Rainbow
+| Red
+| Green
+| Blue
+
+--Now we define the painter function as appears in the Book.
+
+def painter (Γ₀ : Type) (_ : LinearOrderedCommGroupWithZero Γ₀) (v : Valuation ℝ Γ₀) :
+ℝ² → Rainbow
+| X => if v (X 0) < v 1 ∧ v (X 1) < v 1 then Rainbow.Red
+  else if v (X 0) < v (X 1) ∧ v (X 1) ≥ v 1 then Rainbow.Green
+  else Rainbow.Blue
+
+
+
+-- The next two lemmas below basically unravel the definition of the painter function which will
+-- be of use in the proof of the lemma on the boundedness of the determinant.
+
+lemma painted_green1 (Γ₀ : Type) (locg : LinearOrderedCommGroupWithZero Γ₀) (v : Valuation ℝ Γ₀)
+ (X : ℝ²) : painter Γ₀ locg v X = Rainbow.Green → v (X 1) ≥  v (1) := by
+  intro h
+  simp only [painter, Fin.isValue, map_one, ge_iff_le] at h
+  -- I want to get rid of this simp with an unfold but if I do this then split_ifs stops working.
+  split_ifs at h with h1 h2
+  rcases h2 with ⟨_,  q⟩
+  rw [v.map_one]
+  exact q
+
+lemma painted_green2  (Γ₀ : Type) (locg : LinearOrderedCommGroupWithZero Γ₀) (v : Valuation ℝ Γ₀)
+ (X : ℝ²) : painter Γ₀ locg v X = Rainbow.Green → v (X 0) < v (X 1) := by
+  intro h
+  simp only [painter, Fin.isValue, map_one, ge_iff_le] at h
+  split_ifs at h with h1 h2
+  rcases h2 with ⟨p,  _⟩
+  exact p
+
+
+
+-- the next lemma should be a cousin? of push_neg but I couldn't get what was in mathlib to work so
+-- I just did it by hand.
+
+lemma dist_negation_over_and (P Q : Prop): ¬(P ∧ Q) ↔ ¬P ∨ ¬Q := by
```

Comment:
For me the proof  ```by tauto``` works as well. Doesn't really matter, but so you know.

Reply:
Lol we should definitely use by tauto here! I'll update this in the branch version so it'll be there in the next pull request.